WPXMLRPC/WPXMLRPCDataCleaner.m

```diff
@@ -191,30 +193,54 @@ - (NSString *)cleanWithTidyIfPresent:(NSString *)str {
  the length of these extra leading characters.
  Check for a good closing tag, and try to repair a broken closing tag.
  */
-- (NSString *)cleanClosingTagIfNeeded:(NSString *)str
+- (NSString *)cleanClosingTagIfNeeded:(NSString *)str lengthOfCharactersPrecedingPreamble:(NSInteger)length
 {
+    // Clean responses, but not requests (for now).
+    if ([str rangeOfString:@"methodResponse"].location == NSNotFound) {
+        return str;
+    }
+
     // Check for breakage.
     NSString *closingTag = @"</methodResponse>";
     NSRange range = [str rangeOfString:closingTag options:NSBackwardsSearch];
     if (range.location != NSNotFound) {
-        // All is well.
+        // Nothing broken. All is well.
         return str;
     }
 
-    // find the range of the closing params or fault tag
+    // Determine the proper closing tags for the request.
+    NSString *closingTags;
     if ([str rangeOfString:@"<params>"].location == NSNotFound) {
-        range = [str rangeOfString:@"</fault>" options:NSBackwardsSearch];
+        closingTags = @"</value></fault></methodResponse>";
     } else {
-        range = [str rangeOfString:@"</params>" options:NSBackwardsSearch];
+        closingTags = @"</param></params></methodResponse>";
+    }
+
+    // If the length of the content cleaned from before the preamble is greater
+    // than the length of the closing tags, then the xml is too damaged to repair.
+    if (length > closingTags) {
+        // we can't fix this
+        return str;
+    }
+
+    // Find the start of the last end tag
+    range = [str rangeOfString:@"<" options:NSBackwardsSearch];
+    if (range.location == NSNotFound) {
+        // This should never happen, but handle the one in a million case.
+        return str;
     }
 
+    // Find the ending stub
+    NSString *stub = [str substringFromIndex:range.location];
+    range = [closingTags rangeOfString:stub];
```

Comment:
Not sure if this is possible, so please bear with me..

Is it possible str could end in `</param><`?  If so, then the ending stub would be `<`, and the code would duplicate the last `</param>` tag ending the variable str with `</param></param></params></methodResponse>`.

If this was possible, then it could be a better idea to look at the end of `str` for any partial match of `closingTags` directly.